### PR TITLE
Add nuScenes-based lidar examples

### LINKED ISSUE #2510: Add Lidar example
Lidar is very common in AV, and is an oft-requested example to show off.

### PR DESCRIPTION
**Pull request by roym899**
### What
Adds two new examples based on the nuScenes dataset. One example visualizing most of the data (lidar, radar, camera, and 3D bounding boxes) in the nuScenes dataset, and a more minimal example that only visualizes the lidar data in the lidar reference frame.

Lidar example
lidar

nuScenes example
nuscenes

The download module is included in both examples for now; if we have a place so this could be shared across examples, that would be nice.

Closes #2510.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [app.rerun.io](https://app.rerun.io/pr/4407) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4407)
- [Docs preview](https://rerun.io/preview/72f9caff3681fe2cde30f7c0f2ec193ded29831b/docs) 
- [Examples preview](https://rerun.io/preview/72f9caff3681fe2cde30f7c0f2ec193ded29831b/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: examples/python/nuscenes/main.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        current_camera_token = first_radar_token
+        while current_camera_token != "":
+            sample_data = nusc.get("sample_data", current_camera_token)
+            sensor_name = sample_data["channel"]
+            rr.set_time_seconds("timestamp", (sample_data["timestamp"] - start_timestamp) * 1e-6)
+            data_file_path = nusc.dataroot / sample_data["filename"]
+            pointcloud = nuscenes.RadarPointCloud.from_file(str(data_file_path))
+            points = pointcloud.points[:3].T  # shape after transposing: (num_points, 3)
+            point_distances = np.linalg.norm(points, axis=1)
+            point_colors = cmap(norm(point_distances))
```
Comment (nikolausWest) [inline examples/python/nuscenes/main.py:131]: Link to distance based color mapping

File: examples/python/nuscenes/main.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        rr.log("world/anns", rr.Boxes3D(sizes=sizes, centers=centers, rotations=rotations, class_ids=class_ids))
+        current_sample_token = sample["next"]
+
+
+def log_sensor_calibration(sample_data: dict[str, Any], nusc: nuscenes.NuScenes) -> None:
+    """Log sensor calibration (pinhole camera, sensor poses, etc.)."""
+    sensor_name = sample_data["channel"]
+    calibrated_sensor_token = sample_data["calibrated_sensor_token"]
+    calibrated_sensor = nusc.get("calibrated_sensor", calibrated_sensor_token)
+    rotation_xyzw = np.roll(calibrated_sensor["rotation"], shift=-1)
```
Comment (nikolausWest) [inline examples/python/nuscenes/main.py:169]: Add comment on why you do the shift

File: examples/python/nuscenes/main.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            timeless=True,
+        )
+
+
+def main() -> None:
+    parser = argparse.ArgumentParser(description="Visualizes the nuScenes dataset using the Rerun SDK.")
+    parser.add_argument(
+        "--root_dir",
+        type=pathlib.Path,
+        default="dataset",
```
Comment (nikolausWest) [inline examples/python/nuscenes/main.py:196]: Would it make sense for this dataset dir to default to <script dir>/dataset ?

**Review by nikolausWest — APPROVED** [review]
Realized I accidentally left all my comments as "single comment", sorry! Either way, lgtm!

**Comment by nikolausWest** [comment]
Actually, one more question after playing with it: is the nuScenes dataset missing an annotation context?

File: examples/python/lidar/main.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    start_timestamp = first_sample["timestamp"]  # used to start from 0
+    current_lidar_token = first_sample["data"]["LIDAR_TOP"]
+    while current_lidar_token != "":
+        sample_data = nusc.get("sample_data", current_lidar_token)
+
+        data_file_path = nusc.dataroot / sample_data["filename"]
+        pointcloud = nuscenes.LidarPointCloud.from_file(str(data_file_path))
+        points = pointcloud.points[:3].T  # shape after transposing: (num_points, 3)
+        point_distances = np.linalg.norm(points, axis=1)
+        point_colors = cmap(norm(point_distances))
```
[the remark this answers is not in the document]
Comment (roym899) [reply examples/python/lidar/main.py:60]: Done.

File: examples/python/lidar/main.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+def log_nuscenes_lidar(root_dir: pathlib.Path, dataset_version: str, scene_name: str) -> None:
+    nusc = nuscenes.NuScenes(version=dataset_version, dataroot=root_dir, verbose=True)
+
+    scene = next(s for s in nusc.scene if s["name"] == scene_name)
+
+    rr.log("world", rr.ViewCoordinates.RIGHT_HAND_Z_UP, timeless=True)
+
+    first_sample = nusc.get("sample", scene["first_sample_token"])
+    start_timestamp = first_sample["timestamp"]  # used to start from 0
```
[the remark this answers is not in the document]
Comment (roym899) [reply examples/python/lidar/main.py:51]: Removed.

File: examples/python/nuscenes/main.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+                from_parent=False,
+            ),
+        )
+        current_lidar_token = sample_data["next"]
+
+        data_file_path = nusc.dataroot / sample_data["filename"]
+        pointcloud = nuscenes.LidarPointCloud.from_file(str(data_file_path))
+        points = pointcloud.points[:3].T  # shape after transposing: (num_points, 3)
+        point_distances = np.linalg.norm(points, axis=1)
+        point_colors = cmap(norm(point_distances))
```
[the remark this answers is not in the document]
Comment (roym899) [reply examples/python/nuscenes/main.py:102]: Done.

File: examples/python/nuscenes/main.py (hunk 1 above)
Comment (roym899) [reply examples/python/nuscenes/main.py:131]: Done.

File: examples/python/nuscenes/main.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+
+def log_lidar_and_ego_pose(first_lidar_token: str, start_timestamp: numbers.Number, nusc: nuscenes.NuScenes) -> None:
+    """Log lidar data and vehicle pose."""
+    current_lidar_token = first_lidar_token
+
+    while current_lidar_token != "":
+        sample_data = nusc.get("sample_data", current_lidar_token)
+        sensor_name = sample_data["channel"]
+        rr.set_time_seconds("timestamp", (sample_data["timestamp"] - start_timestamp) * 1e-6)
```
[the remark this answers is not in the document]
Comment (roym899) [reply examples/python/nuscenes/main.py:84]: It's in microseconds. Felt easier to go from microseconds to seconds than from microseconds to nanoseconds. But of course both would work.

File: examples/python/nuscenes/main.py (hunk 2 above)
Comment (roym899) [reply examples/python/nuscenes/main.py:169]: Done.

File: examples/python/nuscenes/main.py (hunk 3 above)
Comment (roym899) [reply examples/python/nuscenes/main.py:196]: The examples with extra modules only run when in the script dir any way (same for arkit_scenes and raw_mesh) due to the import of the other module.

**Comment by roym899** [comment]
I just tried adding an annotation context, but the labels are taking a lot of space. Is there any benefit in having the annotation context with just the ids?

**Comment by roym899** [comment]
Hm, that's strange. I guess `pip install nuscenes-devkit` fails for you?

**Comment by emilk** [comment]
Oh yeah, I was `pip` installing the wrong `requirements.txt` file 🤦  Now it seems to be working (downloading dataset…)

File: examples/python/nuscenes/main.py (hunk 3 above)
Comment (roym899) [reply examples/python/nuscenes/main.py]: Fixed it.

**Pull request merged**